### DISCUSSION TITLE
Adding babel plugins for logical assignment and nullish coalescing operator in nuxt 2

### DISCUSSION THREAD

**Discussion by PatMacalister** [nuxt/nuxt#24304] [Questions (Nuxt 2)]
Hi everyone,

I'm currently running into an issue where a [package](https://www.npmjs.com/package/@particle-network/connect) I import is using [logical assignment operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment) as well as [nullish coalescing operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing), which the default babel configuration can't handle.

## Errors

```
ERROR in ./node_modules/@metamask/utils/node_modules/superstruct/dist/index.mjs 18:27
Module parse failed: Unexpected token (18:27)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|         const { path } = failure;
|         const msg = path.length === 0 ? message : `At path: ${path.join('.')} -- ${message}\`;
>         super(explanation ?? msg);
|         if (explanation != null)
|             this.cause = msg;
```

```
ERROR in ./node_modules/@solana/web3.js/lib/index.browser.esm.js 609:27
Module parse failed: Unexpected token (609:27)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|       for (const accountMeta of ix.keys) {
|         const keyMeta = getOrInsertDefault(accountMeta.pubkey);
>         keyMeta.isSigner ||= accountMeta.isSigner;
|         keyMeta.isWritable ||= accountMeta.isWritable;
|       }
```

## Versions

- Node `16.20.0`
- Nuxt `2.15.8`

## What I've tried

After digging through several github issues like [this](https://github.com/nuxt/nuxt/issues/7822) and this [tweet](https://twitter.com/TheAlexLichter/status/1058296116876320769) by Alex I found out the proper to handle these errors is like this: 

Babel provides the following plugins: 
[Babel Logical Assignment Plugin](https://babeljs.io/docs/babel-plugin-transform-logical-assignment-operators)
[Babel Nullish Coalescing Plugin](https://babeljs.io/docs/babel-plugin-transform-nullish-coalescing-operator)

However, the issue persists after installing these packages and adding them to my build config like this: 
```
// nuxt.config.js

// [...]

build: {
    babel: {
      plugins: [
        "@babel/plugin-transform-classes",
        "@babel/plugin-transform-nullish-coalescing-operator",
        "@babel/plugin-transform-logical-assignment-operators",
      ]
    }
  }
```

Since other people don't seem to run into the same issue as me, I'm sure I'm doing something wrong, I just can't figure out what.

## Reproduction

A reproduction of the issue can be found [here](https://github.com/PatMacalister/particle-repro).

The import causing the error can be found at line 9 of the `HelloWorld.vue` component.

Here's my `package.json` for good measure: 
```
{
  "name": "particle-test",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "dev": "nuxt",
    "build": "nuxt build",
    "start": "nuxt start",
    "generate": "nuxt generate"
  },
  "dependencies": {
    "@particle-network/connect": "^1.2.1",
    "core-js": "2",
    "nuxt": "^2.15.8",
    "vue": "^2.7.10",
    "vue-server-renderer": "^2.7.10",
    "vue-template-compiler": "^2.7.10"
  },
  "devDependencies": {
    "@babel/plugin-transform-classes": "^7.23.3",
    "@babel/plugin-transform-logical-assignment-operators": "^7.23.3",
    "@babel/plugin-transform-nullish-coalescing-operator": "^7.23.3",
    "@babel/runtime-corejs2": "^7.23.2"
  }
}
```

**Comment by CARodriguezG** [comment]
In my case, I was trying to add the openai@4.24.1 package into my Nuxt 2 (v2.15.8) project and also having the same issue. In the [link](https://github.com/nuxt/nuxt/issues/7822) you provided there is already an answer by @manniL in this [comment](https://github.com/nuxt/nuxt/issues/7822#issuecomment-667514776).

As he said:

> You don't even need the babel plugins for nullish coalescing and optional chaining therefore because they are already included in the preset-env by now

So, you don't have to install the babel plugins (I don't know if it works for the ALL the babel plugins). What I've done is to add the package into the `transpile` property and it worked:

```
build: {
    transpile: [
        ({ isClient }) => isClient && 'openai' // <-- this is the package name which gives the error
    ]
},
```

You can find more information about the transpile property on the [official Nuxt 2 documentation](https://v2.nuxt.com/docs/configuration-glossary/configuration-build/#transpile).

**Comment by 7ochem** [comment]
Adding this for people who end up here with the same issue as me:

I added Pinia to a Nuxt2 setup (`@pinia/nuxt@0.11.1` with `pinia@3.0.3`) in preparation for migration/upgrade to Nuxt3+Pinia. Then I got this error in the sub dependency `birpc@2.4.0`:

```
ERROR  in ./node_modules/birpc/dist/index.mjs
Module parse failed: Unexpected token (59:60)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|             timeoutId = setTimeout(() => {
|               try {
>                 const handleResult = options.onTimeoutError?.(method, args);
|                 if (handleResult !== true)
|                   throw new Error([birpc] timeout on calling "${method}");
 @ ./node_modules/@vue/devtools-kit/dist/index.js 4870:0-54 6160:14-25 6177:18-34 6195:9-20
 @ ./node_modules/@vue/devtools-api/dist/index.cjs
 @ ./node_modules/pinia/dist/pinia.mjs
```

I used the mentioned solution of adding the `birpc` to the `transpile` build option:
```javascript
  build: {
    transpile: [
      ({ isClient }) => isClient && 'birpc'
    ],
  },
```
